src/ui/steps.py

```
@@ -4,59 +4,97 @@
+        uploaded = st.file_uploader(
             "Select file",
             type=["xlsx", "csv"],
             key="u_file",
-            on_change=_load_uploaded_file,
         )
+        if uploaded:
+            # Track last processed file to avoid redundant reprocessing on every rerun
+            if (
+                st.session_state.get("last_file_processed") != uploaded.name
+                or st.session_state.get("last_file_size") != uploaded.size
+            ):
```

Ruff (0.15.12)

[warning] Use a single `if` statement instead of nested `if` statements

(SIM102)